diff --git a/cpp/src/arrow/chunk_resolver.cc b/cpp/src/arrow/chunk_resolver.cc
index 7fc259f38c..2d9d50b337 100644
--- a/cpp/src/arrow/chunk_resolver.cc
+++ b/cpp/src/arrow/chunk_resolver.cc
@@ -26,7 +26,7 @@
 #include "arrow/array.h"
 #include "arrow/record_batch.h"
 
-namespace arrow {
+namespace arrow::internal {
 
 using util::span;
 
diff --git a/cpp/src/arrow/chunk_resolver.h b/cpp/src/arrow/chunk_resolver.h
index 3d6458167f..75f4d19b1f 100644
--- a/cpp/src/arrow/chunk_resolver.h
+++ b/cpp/src/arrow/chunk_resolver.h
@@ -28,7 +28,7 @@
 #include "arrow/util/macros.h"
 #include "arrow/util/span.h"
 
-namespace arrow {
+namespace arrow::internal {
 
 class ChunkResolver;
 
diff --git a/cpp/src/arrow/compute/api_vector.cc b/cpp/src/arrow/compute/api_vector.cc
index 53ceed1b08..da356ef3fc 100644
--- a/cpp/src/arrow/compute/api_vector.cc
+++ b/cpp/src/arrow/compute/api_vector.cc
@@ -42,13 +42,13 @@ namespace arrow {
 using internal::checked_cast;
 using internal::checked_pointer_cast;
 
-namespace internal {
+namespace internal::vendored {
 
-using compute::DictionaryEncodeOptions;
-using compute::FilterOptions;
+using compute::internal::vendored::DictionaryEncodeOptions;
+using compute::internal::vendored::FilterOptions;
 using compute::NullPlacement;
-using compute::RankOptions;
-using compute::RankQuantileOptions;
+using compute::internal::vendored::RankOptions;
+using compute::internal::vendored::RankQuantileOptions;
 
 template <>
 struct EnumTraits<FilterOptions::NullSelectionBehavior>
@@ -122,8 +122,9 @@ namespace compute {
 // Function options
 
 namespace internal {
+namespace vendored {
 namespace {
-using ::arrow::internal::DataMember;
+using ::arrow::internal::vendored::DataMember;
 static auto kFilterOptionsType = GetFunctionOptionsType<FilterOptions>(
     DataMember("null_selection_behavior", &FilterOptions::null_selection_behavior));
 static auto kTakeOptionsType = GetFunctionOptionsType<TakeOptions>(
@@ -166,67 +167,66 @@ static auto kInversePermutationOptionsType =
 static auto kScatterOptionsType = GetFunctionOptionsType<ScatterOptions>(
     DataMember("max_index", &ScatterOptions::max_index));
 }  // namespace
-}  // namespace internal
 
 FilterOptions::FilterOptions(NullSelectionBehavior null_selection)
-    : FunctionOptions(internal::kFilterOptionsType),
+    : FunctionOptions(kFilterOptionsType),
       null_selection_behavior(null_selection) {}
 constexpr char FilterOptions::kTypeName[];
 
 TakeOptions::TakeOptions(bool boundscheck)
-    : FunctionOptions(internal::kTakeOptionsType), boundscheck(boundscheck) {}
+    : FunctionOptions(kTakeOptionsType), boundscheck(boundscheck) {}
 constexpr char TakeOptions::kTypeName[];
 
 DictionaryEncodeOptions::DictionaryEncodeOptions(NullEncodingBehavior null_encoding)
-    : FunctionOptions(internal::kDictionaryEncodeOptionsType),
+    : FunctionOptions(kDictionaryEncodeOptionsType),
       null_encoding_behavior(null_encoding) {}
 constexpr char DictionaryEncodeOptions::kTypeName[];
 
 RunEndEncodeOptions::RunEndEncodeOptions(std::shared_ptr<DataType> run_end_type)
-    : FunctionOptions(internal::kRunEndEncodeOptionsType),
+    : FunctionOptions(kRunEndEncodeOptionsType),
       run_end_type{std::move(run_end_type)} {}
 
 ArraySortOptions::ArraySortOptions(SortOrder order, NullPlacement null_placement)
-    : FunctionOptions(internal::kArraySortOptionsType),
+    : FunctionOptions(kArraySortOptionsType),
       order(order),
       null_placement(null_placement) {}
 constexpr char ArraySortOptions::kTypeName[];
 
 SortOptions::SortOptions(std::vector<SortKey> sort_keys, NullPlacement null_placement)
-    : FunctionOptions(internal::kSortOptionsType),
+    : FunctionOptions(kSortOptionsType),
       sort_keys(std::move(sort_keys)),
       null_placement(null_placement) {}
 SortOptions::SortOptions(const Ordering& ordering)
-    : FunctionOptions(internal::kSortOptionsType),
+    : FunctionOptions(kSortOptionsType),
       sort_keys(ordering.sort_keys()),
       null_placement(ordering.null_placement()) {}
 constexpr char SortOptions::kTypeName[];
 
 PartitionNthOptions::PartitionNthOptions(int64_t pivot, NullPlacement null_placement)
-    : FunctionOptions(internal::kPartitionNthOptionsType),
+    : FunctionOptions(kPartitionNthOptionsType),
       pivot(pivot),
       null_placement(null_placement) {}
 constexpr char PartitionNthOptions::kTypeName[];
 
 SelectKOptions::SelectKOptions(int64_t k, std::vector<SortKey> sort_keys)
-    : FunctionOptions(internal::kSelectKOptionsType),
+    : FunctionOptions(kSelectKOptionsType),
       k(k),
       sort_keys(std::move(sort_keys)) {}
 constexpr char SelectKOptions::kTypeName[];
 
 CumulativeOptions::CumulativeOptions(bool skip_nulls)
-    : FunctionOptions(internal::kCumulativeOptionsType), skip_nulls(skip_nulls) {}
+    : FunctionOptions(kCumulativeOptionsType), skip_nulls(skip_nulls) {}
 CumulativeOptions::CumulativeOptions(double start, bool skip_nulls)
     : CumulativeOptions(std::make_shared<DoubleScalar>(start), skip_nulls) {}
 CumulativeOptions::CumulativeOptions(std::shared_ptr<Scalar> start, bool skip_nulls)
-    : FunctionOptions(internal::kCumulativeOptionsType),
+    : FunctionOptions(kCumulativeOptionsType),
       start(std::move(start)),
       skip_nulls(skip_nulls) {}
 constexpr char CumulativeOptions::kTypeName[];
 
 RankOptions::RankOptions(std::vector<SortKey> sort_keys, NullPlacement null_placement,
                          RankOptions::Tiebreaker tiebreaker)
-    : FunctionOptions(internal::kRankOptionsType),
+    : FunctionOptions(kRankOptionsType),
       sort_keys(std::move(sort_keys)),
       null_placement(null_placement),
       tiebreaker(tiebreaker) {}
@@ -234,28 +234,28 @@ constexpr char RankOptions::kTypeName[];
 
 RankQuantileOptions::RankQuantileOptions(std::vector<SortKey> sort_keys,
                                          NullPlacement null_placement)
-    : FunctionOptions(internal::kRankQuantileOptionsType),
+    : FunctionOptions(kRankQuantileOptionsType),
       sort_keys(std::move(sort_keys)),
       null_placement(null_placement) {}
 constexpr char RankQuantileOptions::kTypeName[];
 
 PairwiseOptions::PairwiseOptions(int64_t periods)
-    : FunctionOptions(internal::kPairwiseOptionsType), periods(periods) {}
+    : FunctionOptions(kPairwiseOptionsType), periods(periods) {}
 constexpr char PairwiseOptions::kTypeName[];
 
 ListFlattenOptions::ListFlattenOptions(bool recursive)
-    : FunctionOptions(internal::kListFlattenOptionsType), recursive(recursive) {}
+    : FunctionOptions(kListFlattenOptionsType), recursive(recursive) {}
 constexpr char ListFlattenOptions::kTypeName[];
 
 InversePermutationOptions::InversePermutationOptions(
     int64_t max_index, std::shared_ptr<DataType> output_type)
-    : FunctionOptions(internal::kInversePermutationOptionsType),
+    : FunctionOptions(kInversePermutationOptionsType),
       max_index(max_index),
       output_type(std::move(output_type)) {}
 constexpr char InversePermutationOptions::kTypeName[];
 
 ScatterOptions::ScatterOptions(int64_t max_index)
-    : FunctionOptions(internal::kScatterOptionsType), max_index(max_index) {}
+    : FunctionOptions(kScatterOptionsType), max_index(max_index) {}
 constexpr char ScatterOptions::kTypeName[];
 
 namespace internal {
@@ -473,6 +473,7 @@ Result<Datum> Scatter(const Datum& values, const Datum& indices,
                       const ScatterOptions& options, ExecContext* ctx) {
   return CallFunction("scatter", {values, indices}, &options, ctx);
 }
-
+}  // namespace vendored
+}  // namespace internal
 }  // namespace compute
 }  // namespace arrow
diff --git a/cpp/src/arrow/compute/api_vector.h b/cpp/src/arrow/compute/api_vector.h
index 22bb164719..1bcc2362cc 100644
--- a/cpp/src/arrow/compute/api_vector.h
+++ b/cpp/src/arrow/compute/api_vector.h
@@ -30,6 +30,8 @@ namespace compute {
 
 class ExecContext;
 
+namespace internal::vendored {
+
 /// \addtogroup compute-concrete-options
 /// @{
 
@@ -811,5 +813,6 @@ Result<Datum> Scatter(const Datum& values, const Datum& indices,
                       const ScatterOptions& options = ScatterOptions::Defaults(),
                       ExecContext* ctx = NULLPTR);
 
+}  // namespace internal::vendored
 }  // namespace compute
 }  // namespace arrow
diff --git a/cpp/src/arrow/compute/function_internal.h b/cpp/src/arrow/compute/function_internal.h
index 9d8928466b..5f9a607ee9 100644
--- a/cpp/src/arrow/compute/function_internal.h
+++ b/cpp/src/arrow/compute/function_internal.h
@@ -42,6 +42,7 @@ struct StructScalar;
 using ::arrow::internal::checked_cast;
 
 namespace internal {
+namespace vendored {
 template <>
 struct EnumTraits<compute::SortOrder>
     : BasicEnumTraits<compute::SortOrder, compute::SortOrder::Ascending,
@@ -57,13 +58,14 @@ struct EnumTraits<compute::SortOrder>
     return "<INVALID>";
   }
 };
+}  // namespace vendored
 }  // namespace internal
 
 namespace compute {
 namespace internal {
 
-using arrow::internal::EnumTraits;
-using arrow::internal::has_enum_traits;
+using arrow::internal::vendored::EnumTraits;
+using arrow::internal::vendored::has_enum_traits;
 
 template <typename Enum, typename CType = typename std::underlying_type<Enum>::type>
 Result<Enum> ValidateEnumValue(CType raw) {
@@ -656,7 +658,7 @@ template <typename Options, typename... Properties>
 const FunctionOptionsType* GetFunctionOptionsType(const Properties&... properties) {
   static const class OptionsType : public GenericOptionsType {
    public:
-    explicit OptionsType(const arrow::internal::PropertyTuple<Properties...> properties)
+    explicit OptionsType(const arrow::internal::vendored::PropertyTuple<Properties...> properties)
         : properties_(properties) {}
 
     const char* type_name() const override { return Options::kTypeName; }
@@ -694,8 +696,8 @@ const FunctionOptionsType* GetFunctionOptionsType(const Properties&... propertie
     }
 
    private:
-    const arrow::internal::PropertyTuple<Properties...> properties_;
-  } instance(arrow::internal::MakeProperties(properties...));
+    const arrow::internal::vendored::PropertyTuple<Properties...> properties_;
+  } instance(arrow::internal::vendored::MakeProperties(properties...));
   return &instance;
 }
 
diff --git a/cpp/src/arrow/compute/kernels/chunked_internal.cc b/cpp/src/arrow/compute/kernels/chunked_internal.cc
index e72b8e1f5b..dcc72c427c 100644
--- a/cpp/src/arrow/compute/kernels/chunked_internal.cc
+++ b/cpp/src/arrow/compute/kernels/chunked_internal.cc
@@ -22,7 +22,7 @@
 #include "arrow/record_batch.h"
 #include "arrow/util/logging.h"
 
-namespace arrow::compute::internal {
+namespace arrow::compute::internal::vendored {
 
 std::vector<const Array*> GetArrayPointers(const ArrayVector& arrays) {
   std::vector<const Array*> pointers(arrays.size());
diff --git a/cpp/src/arrow/compute/kernels/chunked_internal.h b/cpp/src/arrow/compute/kernels/chunked_internal.h
index 5bc8233016..2884acb541 100644
--- a/cpp/src/arrow/compute/kernels/chunked_internal.h
+++ b/cpp/src/arrow/compute/kernels/chunked_internal.h
@@ -28,7 +28,7 @@
 #include "arrow/compute/kernels/codegen_internal.h"
 #include "arrow/util/span.h"
 
-namespace arrow::compute::internal {
+namespace arrow::compute::internal::vendored {
 
 // The target chunk in a chunked array.
 struct ResolvedChunk {
@@ -77,7 +77,7 @@ struct CompressedChunkLocation {
       : data_((index_in_chunk << kChunkIndexBits) | chunk_index) {}
 
   template <typename IndexType>
-  explicit operator TypedChunkLocation<IndexType>() {
+  explicit operator arrow::internal::TypedChunkLocation<IndexType>() {
     return {static_cast<IndexType>(chunk_index()),
             static_cast<IndexType>(index_in_chunk())};
   }
@@ -90,7 +90,7 @@ static_assert(sizeof(uint64_t) == sizeof(CompressedChunkLocation));
 
 class ChunkedArrayResolver {
  private:
-  ChunkResolver resolver_;
+  ::arrow::internal::ChunkResolver resolver_;
   util::span<const Array* const> chunks_;
   std::vector<const Array*> owned_chunks_;
 
diff --git a/cpp/src/arrow/compute/kernels/codegen_internal.h b/cpp/src/arrow/compute/kernels/codegen_internal.h
index 2a492f581f..0832399be4 100644
--- a/cpp/src/arrow/compute/kernels/codegen_internal.h
+++ b/cpp/src/arrow/compute/kernels/codegen_internal.h
@@ -141,29 +141,6 @@ struct GetViewType<Type, enable_if_t<is_base_binary_type<Type>::value ||
   static T LogicalValue(PhysicalType value) { return value; }
 };
 
-template <>
-struct GetViewType<Decimal32Type> {
-  using T = Decimal32;
-  using PhysicalType = std::string_view;
-
-  static T LogicalValue(PhysicalType value) {
-    return Decimal32(reinterpret_cast<const uint8_t*>(value.data()));
-  }
-
-  static T LogicalValue(T value) { return value; }
-};
-
-template <>
-struct GetViewType<Decimal64Type> {
-  using T = Decimal64;
-  using PhysicalType = std::string_view;
-
-  static T LogicalValue(PhysicalType value) {
-    return Decimal64(reinterpret_cast<const uint8_t*>(value.data()));
-  }
-
-  static T LogicalValue(T value) { return value; }
-};
 
 template <>
 struct GetViewType<Decimal128Type> {
@@ -202,16 +179,6 @@ struct GetOutputType<Type, enable_if_t<is_string_like_type<Type>::value>> {
   using T = std::string;
 };
 
-template <>
-struct GetOutputType<Decimal32Type> {
-  using T = Decimal32;
-};
-
-template <>
-struct GetOutputType<Decimal64Type> {
-  using T = Decimal64;
-};
-
 template <>
 struct GetOutputType<Decimal128Type> {
   using T = Decimal128;
@@ -259,8 +226,7 @@ using enable_if_not_floating_value = enable_if_t<!std::is_floating_point<T>::val
 
 template <typename T, typename R = T>
 using enable_if_decimal_value =
-    enable_if_t<std::is_same<Decimal32, T>::value || std::is_same<Decimal64, T>::value ||
-                    std::is_same<Decimal128, T>::value ||
+    enable_if_t<std::is_same<Decimal128, T>::value ||
                     std::is_same<Decimal256, T>::value,
                 R>;
 
@@ -390,21 +356,6 @@ struct UnboxScalar<Type, enable_if_has_string_view<Type>> {
   }
 };
 
-template <>
-struct UnboxScalar<Decimal32Type> {
-  using T = Decimal32;
-  static const T& Unbox(const Scalar& val) {
-    return checked_cast<const Decimal32Scalar&>(val).value;
-  }
-};
-
-template <>
-struct UnboxScalar<Decimal64Type> {
-  using T = Decimal64;
-  static const T& Unbox(const Scalar& val) {
-    return checked_cast<const Decimal64Scalar&>(val).value;
-  }
-};
 
 template <>
 struct UnboxScalar<Decimal128Type> {
@@ -1170,10 +1121,6 @@ ArrayKernelExec GeneratePhysicalNumeric(detail::GetTypeId get_id) {
 template <template <typename... Args> class Generator, typename... Args>
 ArrayKernelExec GenerateDecimalToDecimal(detail::GetTypeId get_id) {
   switch (get_id.id) {
-    case Type::DECIMAL32:
-      return Generator<Decimal32Type, Args...>::Exec;
-    case Type::DECIMAL64:
-      return Generator<Decimal64Type, Args...>::Exec;
     case Type::DECIMAL128:
       return Generator<Decimal128Type, Args...>::Exec;
     case Type::DECIMAL256:
@@ -1369,10 +1316,6 @@ ArrayKernelExec GenerateTemporal(detail::GetTypeId get_id) {
 template <template <typename...> class Generator, typename Type0, typename... Args>
 ArrayKernelExec GenerateDecimal(detail::GetTypeId get_id) {
   switch (get_id.id) {
-    case Type::DECIMAL32:
-      return Generator<Type0, Decimal32Type, Args...>::Exec;
-    case Type::DECIMAL64:
-      return Generator<Type0, Decimal64Type, Args...>::Exec;
     case Type::DECIMAL128:
       return Generator<Type0, Decimal128Type, Args...>::Exec;
     case Type::DECIMAL256:
diff --git a/cpp/src/arrow/compute/kernels/vector_rank.cc b/cpp/src/arrow/compute/kernels/vector_rank.cc
index d1323c2030..73c050d933 100644
--- a/cpp/src/arrow/compute/kernels/vector_rank.cc
+++ b/cpp/src/arrow/compute/kernels/vector_rank.cc
@@ -23,7 +23,7 @@
 #include "arrow/compute/registry.h"
 #include "arrow/util/math_internal.h"
 
-namespace arrow::compute::internal {
+namespace arrow::compute::internal::vendored {
 
 using ::arrow::util::span;
 
@@ -71,7 +71,7 @@ Result<NullPartitionResult> DoSortAndMarkDuplicate(
   using GetView = GetViewType<ArrowType>;
   using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
 
-  ARROW_ASSIGN_OR_RAISE(auto array_sorter, GetArraySorter(*physical_type));
+  ARROW_ASSIGN_OR_RAISE(auto array_sorter, arrow::compute::internal::GetArraySorter(*physical_type));
 
   ArrayType array(input.data());
   ARROW_ASSIGN_OR_RAISE(auto sorted,
@@ -380,7 +380,7 @@ class RankMetaFunction : public RankMetaFunctionBase<RankMetaFunction> {
   }
 
   RankMetaFunction()
-      : RankMetaFunctionBase("rank", Arity::Unary(), rank_doc, &kDefaultOptions) {}
+      : RankMetaFunctionBase("vendored_rank", Arity::Unary(), rank_doc, &kDefaultOptions) {}
 
   static inline const auto kDefaultOptions = RankOptions::Defaults();
 };
@@ -396,7 +396,7 @@ class RankQuantileMetaFunction : public RankMetaFunctionBase<RankQuantileMetaFun
   static RankerType GetRanker(const RankQuantileOptions& options) { return RankerType(); }
 
   RankQuantileMetaFunction()
-      : RankMetaFunctionBase("rank_quantile", Arity::Unary(), rank_quantile_doc,
+      : RankMetaFunctionBase("vendored_rank_quantile", Arity::Unary(), rank_quantile_doc,
                              &kDefaultOptions) {}
 
   static inline const auto kDefaultOptions = RankQuantileOptions::Defaults();
@@ -413,7 +413,7 @@ class RankNormalMetaFunction : public RankMetaFunctionBase<RankNormalMetaFunctio
   static RankerType GetRanker(const RankQuantileOptions& options) { return RankerType(); }
 
   RankNormalMetaFunction()
-      : RankMetaFunctionBase("rank_normal", Arity::Unary(), rank_normal_doc,
+      : RankMetaFunctionBase("vendored_rank_normal", Arity::Unary(), rank_normal_doc,
                              &kDefaultOptions) {}
 
   static inline const auto kDefaultOptions = RankQuantileOptions::Defaults();
@@ -427,4 +427,4 @@ void RegisterVectorRank(FunctionRegistry* registry) {
   DCHECK_OK(registry->AddFunction(std::make_shared<RankNormalMetaFunction>()));
 }
 
-}  // namespace arrow::compute::internal
+}  // namespace arrow::compute::internal::vendored
diff --git a/cpp/src/arrow/compute/kernels/vector_sort_internal.h b/cpp/src/arrow/compute/kernels/vector_sort_internal.h
index 6288aa26ea..fc03e187c6 100644
--- a/cpp/src/arrow/compute/kernels/vector_sort_internal.h
+++ b/cpp/src/arrow/compute/kernels/vector_sort_internal.h
@@ -31,6 +31,8 @@
 
 namespace arrow::compute::internal {
 
+namespace vendored {
+
 // Visit all physical types for which sorting is implemented.
 #define VISIT_SORTABLE_PHYSICAL_TYPES(VISIT) \
   VISIT(BooleanType)                         \
@@ -456,26 +458,6 @@ using MergeImpl = GenericMergeImpl<uint64_t, NullPartitionResult>;
 using ChunkedMergeImpl =
     GenericMergeImpl<CompressedChunkLocation, ChunkedNullPartitionResult>;
 
-// TODO make this usable if indices are non trivial on input
-// (see ConcreteRecordBatchColumnSorter)
-// `offset` is used when this is called on a chunk of a chunked array
-using ArraySortFunc = std::function<Result<NullPartitionResult>(
-    uint64_t* indices_begin, uint64_t* indices_end, const Array& values, int64_t offset,
-    const ArraySortOptions& options, ExecContext* ctx)>;
-
-Result<ArraySortFunc> GetArraySorter(const DataType& type);
-
-Result<NullPartitionResult> SortChunkedArray(ExecContext* ctx, uint64_t* indices_begin,
-                                             uint64_t* indices_end,
-                                             const ChunkedArray& chunked_array,
-                                             SortOrder sort_order,
-                                             NullPlacement null_placement);
-
-Result<NullPartitionResult> SortChunkedArray(
-    ExecContext* ctx, uint64_t* indices_begin, uint64_t* indices_end,
-    const std::shared_ptr<DataType>& physical_type, const ArrayVector& physical_chunks,
-    SortOrder sort_order, NullPlacement null_placement);
-
 Result<NullPartitionResult> SortStructArray(ExecContext* ctx, uint64_t* indices_begin,
                                             uint64_t* indices_end,
                                             const StructArray& array,
@@ -761,9 +743,9 @@ struct ResolvedTableSortKey {
         order(order),
         null_count(null_count) {}
 
-  using LocationType = ::arrow::ChunkLocation;
+  using LocationType = ::arrow::internal::ChunkLocation;
 
-  ResolvedChunk GetChunk(::arrow::ChunkLocation loc) const {
+  ResolvedChunk GetChunk(::arrow::internal::ChunkLocation loc) const {
     return {chunks[loc.chunk_index], loc.index_in_chunk};
   }
 
@@ -788,7 +770,7 @@ struct ResolvedTableSortKey {
                                   null_count);
     };
 
-    return ::arrow::compute::internal::ResolveSortKeys<ResolvedTableSortKey>(
+    return ::arrow::compute::internal::vendored::ResolveSortKeys<ResolvedTableSortKey>(
         *table.schema(), sort_keys, factory);
   }
 
@@ -813,4 +795,22 @@ inline Result<std::shared_ptr<ArrayData>> MakeMutableFloat64Array(
   return ArrayData::Make(float64(), length, {nullptr, std::move(data)}, /*null_count=*/0);
 }
 
+}  // namespace vendored
+// TODO make this usable if indices are non trivial on input
+// (see ConcreteRecordBatchColumnSorter)
+// `offset` is used when this is called on a chunk of a chunked array
+using ArraySortFunc = std::function<Result<vendored::NullPartitionResult>(
+  uint64_t* indices_begin, uint64_t* indices_end, const Array& values, int64_t offset,
+  const vendored::ArraySortOptions& options, ExecContext* ctx)>;
+Result<ArraySortFunc> GetArraySorter(const DataType& type);
+Result<vendored::NullPartitionResult> SortChunkedArray(ExecContext* ctx, uint64_t* indices_begin,
+  uint64_t* indices_end,
+  const ChunkedArray& chunked_array,
+  SortOrder sort_order,
+  NullPlacement null_placement);
+
+Result<vendored::NullPartitionResult> SortChunkedArray(
+ExecContext* ctx, uint64_t* indices_begin, uint64_t* indices_end,
+const std::shared_ptr<DataType>& physical_type, const ArrayVector& physical_chunks,
+SortOrder sort_order, NullPlacement null_placement);
 }  // namespace arrow::compute::internal
diff --git a/cpp/src/arrow/util/reflection_internal.h b/cpp/src/arrow/util/reflection_internal.h
index 5d281a265f..62cb2cda7b 100644
--- a/cpp/src/arrow/util/reflection_internal.h
+++ b/cpp/src/arrow/util/reflection_internal.h
@@ -26,6 +26,7 @@
 
 namespace arrow {
 namespace internal {
+namespace vendored {
 
 template <typename...>
 struct all_same : std::true_type {};
@@ -131,5 +132,6 @@ struct has_enum_traits : std::false_type {};
 template <typename T>
 struct has_enum_traits<T, void_t<typename EnumTraits<T>::Type>> : std::true_type {};
 
+}  // namespace vendored
 }  // namespace internal
 }  // namespace arrow
